### ISSUE TITLE
azurerm_key_vault / azurerm_key_vault_secret : Error 403 "does not have secrets purge permission on key vault" when deleting secret, but secret gets deleted regardless?

### ISSUE THREAD

**Issue by 9numbernine9** [terraform-providers/terraform-provider-azurerm#10112]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform (and AzureRM Provider) Version

```
Terraform v0.14.4
+ provider registry.terraform.io/hashicorp/azuread v1.1.1
+ provider registry.terraform.io/hashicorp/azurerm v2.42.0
+ provider registry.terraform.io/hashicorp/random v3.0.0
```

### Affected Resource(s)

* `azurerm_key_vault`
* `azurerm_key_vault_secret`

### Terraform Configuration Files

Long-ish sample configuration here: https://gist.github.com/9numbernine9/2d90a2e42758fd2f3deff8334956b8eb

### Debug Output

Trace log of `terraform destroy`
https://gist.githubusercontent.com/9numbernine9/ebafdb65202cb3b255adc22b5fddc9eb/raw/50061e7ebde6c14692f07457cbd44bc49f0da844/trace_destroy_log.txt

### Expected Behaviour

All resources should be successfully destroyed.

### Actual Behaviour

- All resources remain, _except_ `azurerm_key_vault_secret`
- `azurerm_key_vault_secret` is destroyed, **but** `terraform destroy` fails with a `403 Forbidden` error.
- State is left in an unusual, err, state because the local state file believes that the `azurerm_key_vault_secret` still exists but running `az keyvault secret list` shows an empty keyvault.

### Steps to Reproduce

1. `terraform apply` - all resources should be created successfuly.
2. `terraform destroy` - `azurerm_key_vault_secret` fails to be destroyed due to `403 Forbidden`

### Important Factoids

- Tried this with both `azurerm` versions `2.41.0` and `2.42.0` but the behaviour appears to be the same. 

### References

* #9984 - Possibly related?
* [Azure Keyvault soft delete change](https://docs.microsoft.com/en-us/azure/key-vault/general/soft-delete-change) was recently put into effect by Microsoft - possibly also related?

### Notes

Hello! I recently noticed an issue where attempting to `terraform destroy` an `azure_key_vault_secret` results in an unusual situation where Terraform both destroys the secret in the keyvault but _also_ fails due to a `403 Forbidden` error returned from Azure. The errors are always of this form: `Error: purging Secret "debug-admin-username" (Key Vault "https://deletme951bcd23307d9094.vault.azure.net/"): keyvault.BaseClient#PurgeDeletedSecret: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=04b07795-8ddb-461a-bbee-02f9e1bf7b46;oid=7de93dd4-bac0-4b3a-b8d3-1d1c7779d4e7;numgroups=12;iss=https://sts.windows.net/7cc12d39-5552-4b22-80b6-70296cdd3312/' does not have secrets purge permission on key vault 'deletme951bcd23307d9094;location=eastus'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"ForbiddenByPolicy"}`

This puts Terraform into a funky state where subsequent `destroy` operations believe that it still needs to destroy the secret, but if you look in the keyvault in the Azure Portal or use `az keyvault secret list`, the keyvault will show the secret as being removed. 

I've included a sample configuration in a Gist above that shows the minimal setup that I was able to use to reproduce the problem. Note in the sample that I've created a policy on the keyvault that should literally give me every permission available. After creating the secret via Terraform, I've tried deleting it through the Azure Portal and everything seems to work fine - suffice to say that I'm a little confused by the error being returned by Terraform! My second Gist above shows a full run with `TF_LOG=trace terraform destroy` which will hopefully be helpful. 

Let me know if I can provide any further detail for this issue!

**Comment by gitege** [comment]
Hi, I have seen the same behaviour as you. Soft delete is now enforced on Key Vault so the secrets are being deleted but they are still there and have to be purged in order for Terraform to completely destroy. Looking at the code the purge is called within the delete but fails because purge protection is enabled in your configuration. Rerunning is actually successful because the secret is soft deleted so Terraform doesn't attempt to purge second time round.

Changing purge_protection_enabled  to false worked in my case when creating a new key vault, you can't disable purge protection after it is set however
https://docs.microsoft.com/en-us/azure/key-vault/general/key-vault-recovery?tabs=azure-portal

**Comment by esciara** [comment]
I have the same error when doing a `terraform destroy`. 

When doing a second `terraform destroy`, I get the error:

`Error: deleting Secret "XXX" (Key Vault "https://XXX.vault.azure.net/"): keyvault.BaseClient#DeleteSecret: Failure responding to request: StatusCode=403 -- Original Error: autorest/azure: Service returned an error. Status=403 Code="Forbidden" Message="The user, group or application 'appid=XXX;oid=XXX=2;iss=https://sts.windows.net/XXX/' does not have secrets delete permission on key vault 'base-frct-dev-kv;location=francecentral'. For help resolving this issue, please see https://go.microsoft.com/fwlink/?linkid=2125287" InnerError={"code":"AccessDenied"}`

which is due to the fact that the policy has been deleted as part of the destroy, even though I had a `depends_on` flag on the `azurerm_key_vault_secret` resource to make it depends on my `azurerm_key_vault_access_policy`.

After doing a `terraform refresh`, another `terraform destroy` indicates:
```
Destroy complete! Resources: 0 destroyed.
```

I am using `terraform` version `0.14.4` and `azurerm` version `2.42.0`. The [release notes for `2.43.0`](https://github.com/terraform-providers/terraform-provider-azurerm/releases/tag/v2.43.0) do not indicate any bug correction around key vaults.

**Comment by 9numbernine9** [comment]
@gitege Thanks for your insight! When I was creating that sample configuration, I copied the `azurerm_key_vault` configuration from a configuration where I'm using the keyvault to store both some `azurerm_key_vault_secret`s _and_ an `azurerm_key_vault_key` that's being used for disk encryption -- the key used for disk encryption _must_ be placed in a keyvault that has purge protection enabled. 

With that said, it seems like it might not be possible to have both `purge_soft_delete_on_destroy = true` and a keyvault with `purge_protection_enabled = true` right now. Although I wanted purge keys/secrets on destroy, it might be nice if Terraform wouldn't try to do that if I've enabled purge protection on the affected keyvault that contains those keys/secrets. 😞
